docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,480 @@
+
+  Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+  // ControlPlane is a reference to a local struct that holds the details
+  // for provisioning the Control Plane for the Cluster.
+  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+  // Worker describes the worker nodes for the cluster.
+  // It is a collection of node types which can be used to create
+  // the worker nodes of the cluster.
+  // +optional
+  Worker WorkerClass `json:"worker,omitempty"`
+
+  // Infrastructure is a reference to a provider-specific template that holds
+  // the details for provisioning infrastructure specific cluster
+  // for the underlying provider.
+  // The underlying provider is responsible for the implementation
+  // of the template to an infrastructure cluster.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes which pronounce
+//
+type WorkerClass struct {
+  // Deployments is a list of machine deployment types that can be used to create
+  // a set of worker plane nodes.
+  Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+  // Class denotes a type of worker node present in the cluster
+  Class string `json:"class,omitempty"`
+
+
+  // Template is a local struct containing a collection of templates for creation of
+  // MachineDeployment objects representing a set of worker nodes.
+  Template TemplateClass `json:"template"`
+}
+
+type TemplateClass struct {
+  Metadata ObjectMeta `json:"metadata,omitempty"`
+
+  // Bootstrap contains the bootstrap template references to be used
+  // for the creation of worker Machines.
+  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+  // Infrastructure contains the infrastructure template references to be used
+  // for the creation of worker Machines.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+type LocalObjectTemplate struct {
+  // Ref is a required reference to a custom resource
+  // offered by a bootstrap provider.
+  Ref corev1.ObjectReference `json:"ref"`
+}
+```
+
+#### Modification to existing API Types
+##### Cluster
+1.  Add `Cluster.Spec.Topology` defined as
+    ```golang
+    // This encapsulates the managed topology for the cluster.
+    // +optional
+    Topology *ManagedTopology `json:"topology,omitempty"`
+    ```
+1.  The `ManagedTopology` object has the following definition:
+    ```golang
+    // ManagedTopology encapsulates the information of the managed resources.
+    type ManagedTopology struct {
+      // The name of the ClusterClass object to create the managed topology.
+      Class string `json:"class,omitempty"`
```

Comment:
Does a managed topology have a status?
If not how do we know e.g in which version the cluster is?
Did it reconcile successfully last time spec.Version was changed?
What is successfully, is it the controlPlane upgraded or does it include all workers?

If we don't want to answer all this questions about upgrades we can may be drop the version field for now and add in a follow up with fleshed details or if we want to keep it we should add an upgrade caveat section or so capturing all these unknowns.

Reply:
These all are valid concerns, longer term we can probably mirror some of these fields as a status field, although given that these APIs are not set in stone yet we can add conditions over time to signal.

A lot of these questions can be answered once the technical implementation is completed, given that we need to understand how the system works to properly signal successful version upgrades and possible caveats. We can amend the proposal along the way, although I wouldn't necessarily want to remove the version field now, but rather add more informational bits once we're ready to commit to some specific behaviors within the managed topology functionality.

Reply:
@enxebre Is this an acceptable path to move forward and retrofit the proposal with more details around this part?